File: go/p39.go
package main

import "sort"

func combinationSum(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	resultList := make([][]int, 0)
	dfs(candidates, target, 0, make([]int, 0), &resultList)
	return resultList
}

func dfs(candidates []int, target int, start int, valueList []int, resultList *[][]int) {
	length := len(candidates)
	if target == 0 {
		tmpCurComb := make([]int, len(valueList))
		copy(tmpCurComb, valueList)
		*resultList = append(*resultList, tmpCurComb)
		return
	}
	for i := start; i < length; i++ {
		if target < candidates[i] {
			break
		}
		dfs(candidates, target-candidates[i], i, append(valueList, candidates[i]), resultList)
	}
}

func main() {
	resultList := combinationSum([]int{2, 3, 7}, 18)
	for _, s := range resultList {
		for _, v := range s {
			print(v)
		}
		println()
	}
}

File: python/p39.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates = sorted(candidates)
        Solution.ret = []
        self.dfs(candidates, target, 0, [])
        return Solution.ret

    def dfs(self, candidates, target, start, value_list):
        length = len(candidates)
        if target == 0:
            Solution.ret.append(value_list)
        for i in range(start, length):
            if target < candidates[i]:
                return
            self.dfs(candidates, target - candidates[i], i, value_list + [candidates[i]])

if __name__ == '__main__':
    print(Solution().combinationSum([8, 7, 4, 3], 11))